x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/RandomizedAssignmentRounding.java

```diff
@@ -0,0 +1,419 @@
+                    }
+                }
+            }
+
+            for (Model m : models.stream().sorted(Comparator.comparingDouble(this::remainingModelOrder)).toList()) {
+                if (resourceTracker.remainingModelAllocations.get(m) > 0) {
+                    for (Node n : nodes.stream()
+                        .sorted(
+                            Comparator.comparingDouble(
+                                n -> remainingNodeOrder(
+                                    n,
+                                    m,
+                                    resourceTracker.remainingNodeCores.get(n),
+                                    resourceTracker.remainingNodeMemory.get(n),
+                                    resourceTracker.remainingModelAllocations.get(m)
+                                )
+                            )
+                        )
+                        .toList()) {
+
+                        Tuple<Model, Node> assignment = Tuple.tuple(m, n);
+                        if (resourceTracker.remainingNodeMemory.get(n) >= m.memoryBytes()
+                            && resourceTracker.remainingNodeCores.get(n) >= m.threadsPerAllocation()
+                            && resultAllocations.get(assignment) == 0) {
+                            int assigningAllocations = Math.min(
+                                resourceTracker.remainingNodeCores.get(n) / m.threadsPerAllocation(),
+                                resourceTracker.remainingModelAllocations.get(m)
+                            );
+                            resourceTracker.assign(m, n, assigningAllocations);
+                            resultAllocations.put(assignment, assigningAllocations);
+                            if (resourceTracker.remainingModelAllocations.get(m) == 0) {
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+            return resultAllocations;
+        }
+
+        private double remainingNodeOrder(
+            Node n,
+            Model m,
+            int remainingNodeCores,
+            long remainingNodeMemory,
+            int remainingModelAllocations
+        ) {
+            // as java.lang.Math#abs(long) is a forbidden API, we compute an abs manually
+            int distanceOfRemainingCoresToModelThreads = remainingNodeCores - remainingModelAllocations * m.threadsPerAllocation();
+            if (distanceOfRemainingCoresToModelThreads < 0) {
+                distanceOfRemainingCoresToModelThreads = -distanceOfRemainingCoresToModelThreads;
+            }
+            return (m.currentAllocationByNodeId().containsKey(n.id()) ? 0 : 1) + (remainingNodeCores <= remainingModelAllocations * m
+                .threadsPerAllocation() ? 0 : 0.5) + (0.01 * distanceOfRemainingCoresToModelThreads) + (0.01 * remainingNodeMemory);
+        }
+    }
+
+    private static class ResourceTracker {
```

Comment:
why not a record? Because you want things `final`?

Reply:
Given its constructor takes different arguments than its members, it wouldn't gain us much to make it a record.